presto-memory-context/src/main/java/com/facebook/presto/memory/context/RootAggregatedMemoryContext.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory.context;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import static java.util.Objects.requireNonNull;
+
+class RootAggregatedMemoryContext
+        extends AggregatedMemoryContext
+{
+    private final MemoryReservationHandler reservationHandler;
+    private final long guaranteedMemory;
+
+    RootAggregatedMemoryContext(MemoryReservationHandler reservationHandler, long guaranteedMemory)
+    {
+        this.reservationHandler = requireNonNull(reservationHandler, "reservationHandler is null");
+        this.guaranteedMemory = guaranteedMemory;
+    }
+
+    synchronized ListenableFuture<?> updateBytes(long bytes)
+    {
+        ListenableFuture<?> future = reservationHandler.reserveMemory(bytes);
+        addBytes(bytes);
+        // make sure we never block queries below guaranteedMemory
+        if (getBytes() < guaranteedMemory) {
+            future = NOT_BLOCKED;
+        }
+        return future;
+    }
+
+    synchronized boolean tryUpdateBytes(long delta)
+    {
+        if (reservationHandler.tryReserveMemory(delta)) {
+            addBytes(delta);
+            return true;
+        }
+        return false;
```

Comment:
This means, while we're under guaranteed memory, we can still return `false` from `tryUpdateBytes` even though call to `updateBytes` would succeed (returning `NOT_BLOCKED`)

Reply:
Yes, which is the current behavior, please see the usage of `QueryContext.GUARANTEED_MEMORY`. That field is added with https://github.com/prestodb/presto/commit/2151eb8db40600ccf3a624eff9a4b61bc8abf07a and `tryReserveMemory` method existed back then and it wan't updated, so I am assuming this is the correct/intended behavior.

Reply:
"current behavior" doesn't imply "correct behavior"/"intended behavior", we're only humans.  However, I can understand that retaining behavior is a safe bet.

Reply:
I agree with you. I am not saying current **implies** correct, I said I am **assuming** it's correct.